src/System.CommandLine.Tests/GetValueByNameParserTests.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using FluentAssertions;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.CommandLine.Tests
+{
+    public class GetValueByNameParserTests : ParserTests
+    {
+        public GetValueByNameParserTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        protected override T GetValue<T>(ParseResult parseResult, Option<T> option)
+            => parseResult.GetValue<T>(option.Name);
+
+        protected override T GetValue<T>(ParseResult parseResult, Argument<T> argument)
+            => parseResult.GetValue<T>(argument.Name);
+
+        [Fact]
+        public void In_case_of_argument_name_conflict_the_value_which_belongs_to_the_last_parsed_command_is_returned()
```

Comment:
A few additional test cases come to mind:

* What happens when an option and an argument name are the same...
  * at the same level of the tree?
  * at different levels of the tree?
* What happens when symbols with the same name exist at different levels of the tree and the inner one is not provided? Does it fall back to the outer one's value?
  * Does this vary depending on whether the inner one has a default value defined?
* How do customer parsers behave when...
  * ...parsing an inner symbol `-x` and looking up a value for `-x` when there's an outer symbol `-x`? 
  * ...parsing an outer symbol `-x` and looking up a value for `-x` when there's an inner symbol `-x`?
  * ...parsing symbol `-a` and looking up a value for symbol `-x`...
    * ...which is above it in the tree
    * ...which is below it in the tree

I have opinions on what some of these behaviors should be but I'll wait so I don't anchor anyone.

Reply:
> * What happens when an option and an argument name are the same at the same level of the tree?
  
We should throw.
  
> * What happens when an option and an argument name are the same at different levels of the tree?

The new methods provides only values for the parsed command (`parseResult.CommandResult`), not for the entire symbol tree.

> * What happens when symbols with the same name exist at different levels of the tree and the inner one is not provided? Does it fall back to the outer one's value?

The outer value will never be returned, only inner if provided or has default.

> How do customer parsers behave when...

Such scenario is not supported by design (nobody asked for it), the method is exposed only for `ParseResult`, not for `SymbolResult`. It simplifies the design a lot (the cache can be populated just once).